components/Buttons/src/ColorThemer/MDCButtonStyler.h

```diff
@@ -0,0 +1,52 @@
+/*
+ Copyright 2017-present the Material Components for iOS authors. All Rights Reserved.
+
+ Licensed under the Apache License, Version 2.0 (the "License");
+ you may not use this file except in compliance with the License.
+ You may obtain a copy of the License at
+
+ http://www.apache.org/licenses/LICENSE-2.0
+
+ Unless required by applicable law or agreed to in writing, software
+ distributed under the License is distributed on an "AS IS" BASIS,
+ WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ See the License for the specific language governing permissions and
+ limitations under the License.
+ */
+
+#import <Foundation/Foundation.h>
+
+@class MDCButton;
+
+/**
+ MDCButtonStyler is a utility class providing methods to style MDCButtons to match different
+ Material Design button styles. The focus is on the shape, elevation, and size of the button.
+ Colors, fonts, and content are left unchanged.
+
+ @Note Some methods change properties not proxied by UIAppearance.
+ */
+@interface MDCButtonStyler : NSObject
```

Comment:
Is this intentionally not being named MDCButtonThemer?

Reply:
It is, but I'm open to new names. So far we have:
* Themer - Changes colors of components
* Mutators - Read properties and then modifies properties.

I thought that since this was a pretty different modification of the button, it might deserve a new name. Many (most?) of these stylers do not support UIAppearance because they will be editing `imageEdgeInsets` and `titleEdgeInsets` to adhere to the Material Design guidelines for that button.  Although the Material guidelines refer to these as "types" of buttons, `MDCButtonTyper` just sounded weird. :-\

Reply:
It seems like we may be trending towards something that would support a functional pattern of configuration quite well. Consider the following (taken from your other comment, @romoore):

- Transform into one of the Material button "types"
- Update colors to match branding
- Mutate text/imagery to ensure accessibility conformance
- Perform mutation X

A functional way to write this in Swift would be:

```swift
let button = MDCButton()
button.mutate(with: [
  MDCButtonAsFloatingButton,
  MyAppBrandedColors,
  MDFTextAccessibility,
  MutationX
])
```

Because we can't easily store blocks in arrays in Objective-C, we'd implement a protocol-based API:

```objc
@protocol MDCButtonMutator
- (void)mutate:(nonnull MDCButton *)button;
@end

@interface MDCButton (Mutation)

- (void)mutateWith:(nonnull NSArray<id<MDCButtonMutator>> *)mutators;

@end
```

Then anybody could implement a mutator at will:

```objc
@implementation MDCButtonAsFloatingButton

- (void)mutate:(MDCButton *)button {

}

@end

MDCButton *button = [[MDCButton alloc] init];
[button mutateWith:@[[[MDCButtonAsFloatingButton alloc] init]]];
```

It would then become a common pattern for client teams to create and store their mutators in this fashion so that order and consistency can be encouraged.

This is similar to our already existing mutator shapes, but it formalizes the pattern so that we can unify on a single name across all of these data types (and hopefully avoid introducing a 3rd, 4th, ....etc data type as we go along).

Reply:
If we wanted to get really daring, we could provide stateful mutators that would be invoked after certain events happen on the instance. E.g. "every time the text changes, run these mutators":

```swift
button.mutateOnTextChange = [
  MDFTextAccessibility()
]
```

Reply:
Huzzah for the [visitor pattern](https://en.wikipedia.org/wiki/Visitor_pattern)!

Reply:
Yussss

Reply:
If we decide to go with the visitor pattern approach, it may be beneficial to split this class up into distinct types for each type of mutation.

Reply:
Interesting approach. I seem to recall a lot of discussion about how to implement Themers.  @jgunaratne probably has the most insight.

Short version:
Themers adjust the visual rendering of a component.  (Not just Colors, those are just the first pass of themers implemented)
Themers should be able to affect an instance or a UIAppearance proxy.
Themers should be stateless.


I would expect this to be named MDCButtonTypeThemer or MDCButtonStyleThemer.

Reply:
Ok, then should I rename these to Mutator so I can edit non-UIAppearance properties?

Reply:
Are the non-UIAppearance properties the insets?  Can we expose them in MDCButton?  Should we follow the pattern we use in TextFields and make them class properties?

Reply:
They are, and we also need to force a "sizeToFit:" pass while mutating the button.  We need to ensure it has the correct height/width when updating the layer's cornerRadius value.

Reply:
re"sizeToFit": UIAppearance compatibility is a nice to have but not required.

If we want to fit this in closer to the Mutator pattern I would have a class for each style each with the same method signature to convert a button to that style. (Regular | Mini | extended).